indefinitepagerindicator/src/main/java/com/rbrooks/indefinitepagerindicator/IndefinitePagerIndicator.kt

```diff
@@ -248,41 +255,71 @@ class IndefinitePagerIndicator @JvmOverloads constructor(
     /**
      * Attach a RecyclerView to the Pager Indicator.
      *
-     * If ViewPager previously attached, remove this class (page change listener) and set to null.
-     * If other RecyclerView previously attached, remove internal scroll listener.
+     * Any previously attached source will be detached.
      */
     fun attachToRecyclerView(recyclerView: RecyclerView?) {
-        viewPager?.removeOnPageChangeListener(this)
-        viewPager = null
-
-        this.recyclerView?.removeOnScrollListener(internalRecyclerScrollListener)
+        detachSources()
 
         this.recyclerView = recyclerView
-        internalRecyclerScrollListener = InternalRecyclerScrollListener()
-        this.recyclerView?.addOnScrollListener(internalRecyclerScrollListener)
+
+        InternalRecyclerScrollListener().let { newScrollListener ->
+            internalRecyclerScrollListener = newScrollListener
+            this.recyclerView?.addOnScrollListener(newScrollListener)
+        }
     }
 
     /**
      * Attach a ViewPager to the Pager Indicator.
      *
-     * If RecyclerView previously attached, scroll listener will be removed and RV set to null.
-     * If other ViewPager previously attached, remove reference to this class (page change listener).
+     * Any previously attached source will be detached.
      */
     fun attachToViewPager(viewPager: ViewPager?) {
-        recyclerView?.removeOnScrollListener(internalRecyclerScrollListener)
-        recyclerView = null
-
-        this.viewPager?.removeOnPageChangeListener(this)
+        detachSources()
 
         this.viewPager = viewPager
         this.viewPager?.addOnPageChangeListener(this)
 
         selectedItemPosition = viewPager?.currentItem!!
     }
 
+    /**
+     * Attach a ViewPager2 to the Pager Indicator.
+     *
+     * Any previously attached source will be detached.
+     */
+    fun attachToViewPager2(viewPager2: ViewPager2) {
+        detachSources()
+
+        this.viewPager2 = viewPager2
+
+        InternalPageChangeCallback().let {
+            internalPageChangeCallback = it
+            this.viewPager2?.registerOnPageChangeCallback(it)
+        }
+
+        selectedItemPosition = this.viewPager2?.currentItem!!
+    }
+
+    private fun detachSources() {
+        internalRecyclerScrollListener?.let {
```

Comment:
I liked this method! However, what do you think about renaming it to something more specific like **removeAllListeners**?

Reply:
Besides removing listeners it also removes references to all sources, so I picked a less specific name that includes everything it does. Is something like `removeAllSources` better?

Reply:
ok, fair. That sounds better, yes.